nomad/resource_namespace.go

```diff
@@ -0,0 +1,144 @@
+				Optional:    true,
+				Type:        schema.TypeString,
+			},
+
+			"quota": {
+				Description: "Quota to set for this namespace.",
+				Optional:    true,
+				Type:        schema.TypeString,
+			},
+		},
+	}
+}
+
+func resourceNamespaceCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	namespace := api.Namespace{
+		Name:        d.Get("name").(string),
+		Description: d.Get("description").(string),
+		Quota:       d.Get("quota").(string),
+	}
+
+	// upsert our namespace
+	log.Printf("[DEBUG] Creating namespace %q", namespace.Name)
+	_, err := client.Namespaces().Register(&namespace, nil)
+	if err != nil {
+		return fmt.Errorf("error inserting namespace %q: %s", namespace.Name, err.Error())
+	}
+	log.Printf("[DEBUG] Created namespace %q", namespace.Name)
+	d.SetId(namespace.Name)
+
+	return resourceNamespaceRead(d, meta)
+}
+
+func resourceNamespaceUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	namespace := api.Namespace{
+		Name:        d.Get("name").(string),
+		Description: d.Get("description").(string),
+		Quota:       d.Get("quota").(string),
+	}
+
+	// upsert our namespace
+	log.Printf("[DEBUG] Updating namespace %q", namespace.Name)
+	_, err := client.Namespaces().Register(&namespace, nil)
+	if err != nil {
+		return fmt.Errorf("error inserting namespace %q: %s", namespace.Name, err.Error())
+	}
+	log.Printf("[DEBUG] Updated namespace %q", namespace.Name)
+
+	return resourceNamespaceRead(d, meta)
+}
+
+func resourceNamespaceDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	name := d.Id()
+
+	// delete the namespace
+	log.Printf("[DEBUG] Deleting namespace %q", name)
+	_, err := client.Namespaces().Delete(name, nil)
+	if err != nil {
+		return fmt.Errorf("error deleting namespace %q: %s", name, err.Error())
+	}
+	log.Printf("[DEBUG] Deleted namespace %q", name)
+
+	return nil
+}
+
+func resourceNamespaceRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	name := d.Id()
+
+	// retrieve the namespace
+	log.Printf("[DEBUG] Reading namespace %q", name)
+	namespace, _, err := client.Namespaces().Info(name, nil)
+	if err != nil {
+		// we have Exists, so no need to handle 404
+		return fmt.Errorf("error reading namespace %q: %s", name, err.Error())
+	}
+	log.Printf("[DEBUG] Read namespace %q", name)
+
+	d.Set("name", namespace.Name)
+	d.Set("description", namespace.Description)
+	d.Set("quota", namespace.Quota)
+
+	return nil
+}
+
+func resourceNamespaceExists(d *schema.ResourceData, meta interface{}) (bool, error) {
+	client := meta.(*api.Client)
+
+	name := d.Id()
+	log.Printf("[DEBUG] Checking if namespace %q exists", name)
+	resp, _, err := client.Namespaces().Info(name, nil)
+	if err != nil {
+		// As of Nomad 0.4.1, the API client returns an error for 404
+		// rather than a nil result, so we must check this way.
+		if strings.Contains(err.Error(), "404") {
```

Comment:
I'm assuming that means the status code also isn't a 404?

Reply:
It probably is, unfortunately the client library doesn't expose that to me :/

Reply:
Please link to https://github.com/hashicorp/nomad/issues/1849 in this comment so we can follow up later more easily.